### PR TITLE
Add quic patches to 3.1.0

### PR DESCRIPTION
**Pull request by wbl**
Pretty sure I got all the version numbers and such in the rebase, but entirely possible I missed a few.

### CODE REVIEW

**Comment by nibanks** [comment]
From my testing so far, perf looks good. I just kicked off our full MsQuic CI automation [here](https://github.com/microsoft/msquic/pull/3511) to fully test this.

**Review by nibanks — APPROVED** [review]
I did not look extensively through the code changes, but based on the MsQuic tests all passing, and perf looking pretty good, I'm approving this.

**Review by richsalz — APPROVED** [review]
why is fips-checksums.yml in this patch?  shrug.  don't care much.  approved.

File: ssl/tls13_enc.c [hunk 1]
```diff
@@ -385,24 +385,231 @@ static int derive_secret_key_and_iv(SSL *s, int sending, const EVP_MD *md,
...
+#else
+static const unsigned char client_early_traffic[] = "c e traffic";
+static const unsigned char client_handshake_traffic[] = "c hs traffic";
+static const unsigned char client_application_traffic[] = "c ap traffic";
+static const unsigned char server_handshake_traffic[] = "s hs traffic";
+static const unsigned char server_application_traffic[] = "s ap traffic";
+static const unsigned char exporter_master_secret[] = "exp master";
+static const unsigned char resumption_master_secret[] = "res master";
+static const unsigned char early_exporter_master_secret[] = "e exp master";
+#endif
```
Comment (tmshort) [inline ssl/tls13_enc.c:406]: A change in 3.1.0 made the `#ifdef CHARSET_EBCDIC` unnecessary. This _could_ be reduced down to the equivalent of lines 390~405 in this file.

**Review by tmshort — CHANGES_REQUESTED** [review]
Just need to add a commit to fix up the version. See https://github.com/quictls/openssl/commit/183e548d93a6aeedab0fd6cb737923a93f807b83

File: README.md [hunk 2]
```diff
@@ -1,224 +1,113 @@
...
-  - creation of X.509 certificates, CSRs and CRLs
-  - calculation of message digests
-  - encryption and decryption
-  - SSL/TLS client and server tests
-  - handling of S/MIME signed or encrypted mail
-  - and more...
+This is a fork of [OpenSSL](https://www.openssl.org) to enable QUIC. In addition
+to the website, the official source distribution is at
+<https://github.com/openssl/openssl>. The OpenSSL `README` can be found at
+[README-OpenSSL.md](https://github.com/quictls/openssl/blob/openssl-3.0.1%2Bquic/README-OpenSSL.md)
```
[the remark this answers is not in the document]
Comment (tmshort) [reply README.md:7]: I see that you probably did not include the last couple README fixup commits.

**Comment by tmshort** [comment]
I did do the:
```
diff <(git show -U0 $commit1) <(git show -U0 $commit2)
```
Command, and everything generally looks good. Other than the two things noted above, all I saw were context and whitespace differences.

File: ssl/tls13_enc.c (hunk 1 above)
Comment (tmshort) [reply ssl/tls13_enc.c:406]: This is mostly a nit, since the code is still correct.

**Comment by tmshort** [comment]
This is the fuzzer error:
```
clang-15: error: unsupported option '--with-fuzzer-lib=/usr/lib/libFuzzingEngine'
```
Looks to be an environment error; I wonder if it's because we don't have an up-to-date `master` branch or something?

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (wbl) [reply README.md:7]: Yeah, I think I tried to be too clever about combining them. Right thing at this point is probably to not be clever about it and cherrypick them to this.

**Comment by wbl** [comment]
I'll go fix the README issue and adjust the other commit Monday and push an updated one.

File: ssl/tls13_enc.c (hunk 1 above)
Comment (tmshort) [reply ssl/tls13_enc.c:406]: No need to make a change. But, if you were to make a change, copy the deleted code from the original file at 390~405, into the position above. Basically, you can get rid of the `#ifdef CHARSET_EBCDIC` conditional that way.

**Comment by wbl** [comment]
I've added in the README commits and further bumped the version in the README. I'll pass on messing with the EBDIC stuff since Todd doesn't think there is a need and I'm not quite sure I understand it.

**Review by nibanks — APPROVED** [review]

**Comment by pheiduck** [comment]
> I wonder if it's because we don't have an up-to-date `master` branch or something?
@tmshort Yep, This branch uses clang-15 and `master` clang-12 where the error does not occur.

**Comment by wbl** [comment]
I was hoping @tmshort would be able to look again to confirm I did the README right, or @kaduk be able to take a look. If you think its good to merge despite that, go ahead: I'm not quite sure what our norms are.

**Comment by nibanks** [comment]
> I was hoping @tmshort would be able to look again to confirm I did the README right, or @kaduk be able to take a look. If you think its good to merge despite that, go ahead: I'm not quite sure what our norms are.

I would like one of them to take a look as well if possible.

**Comment by nibanks** [comment]
@tmshort or @kaduk would you be able to take a quick look? Thanks!

**Comment by tmshort** [comment]
Sorry; I'm not getting notifications (or they are being getting lost) for this repo.

**Review by tmshort — APPROVED** [review]
Apologies for not looking at this in a timely manner. I have to figure out where my github notifications are going... I will try not to have this happen next time!

**Comment by tmshort** [comment]
Still not worried about Fuzz. I will push the latest OpenSSL master branch, even though this repo won't use it. We will likely have to keep that one up to date as well.

**Pull request merged**